### PR TITLE
[Fixes #4087] Add AddViewComponentsAsServices() and ServiceBasedViewComponentActivator

### LINKED ISSUE #4087: No equivalent of AddControllersAsServices() for VC's and TagHelpers
If you have an external assembly, contianing a Controller, and a ViewComponent, you can register the controller with MVC by calling `AddControllersAsServices()`

However there is no equivalent for the ViewComponent.
The default behaviour seems to be if you have a project reference to the assembly then the VC is picked up, but if you don't have a project reference to it (i.e `ILibraryManager` is not aware of it) - because for example, it's a standalone "plugin" assembly, then VC's are not detected. 

Could you add a method similar to `AddControllersAsServices()` but for VC's?

Or perhaps consider adding some sort of "IncludeMvcServicesFromAssemblies()` method, that will make sure the assemblies are treated as candidates for all types of services that MVC looks for - i.e Controllers, VC's, and whatever else.

### PR DESCRIPTION
**Pull request by javiercn**
Added view components as a feature and wired them up through the stack

### CODE REVIEW

**Comment by rynowak** [comment]
:watch: design looks good

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/DependencyInjection/MvcViewFeaturesMvcCoreBuilderExtensions.cs [hunk 1]
```diff
@@ -26,10 +28,19 @@ public static IMvcCoreBuilder AddViews(this IMvcCoreBuilder builder)
             }
 
             builder.AddDataAnnotations();
+            AddViewApplicationPartsProviders(builder.Manager);
             AddViewServices(builder.Services);
             return builder;
         }
 
+        private static void AddViewApplicationPartsProviders(ApplicationPartManager manager)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/DependencyInjection/MvcViewFeaturesMvcCoreBuilderExtensions.cs]: View -> ViewComponent

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    public class ServiceBasedViewComponentActivator : IViewComponentActivator
+    {
+        public object Create(ViewComponentContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            var controllerType = context.ViewComponentDescriptor.TypeInfo.AsType();
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs]: ???

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+            return context.ViewContext.HttpContext.RequestServices.GetRequiredService(controllerType);
+        }
+
+        public virtual void Release(ViewComponentContext context, object viewComponent)
+        {
+        }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs]: document that dispose is handled by service provider

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentsFeature.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Mvc.ViewComponents
+{
+    public class ViewComponentsFeature
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentsFeature.cs]: Let's make these names singular

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentsFeature.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentsFeature.cs]: license

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs [hunk 6]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs]: license

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentsFeatureProvider.cs [hunk 7]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ViewComponentsFeatureProvider.cs]: license

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs (hunk 3 above)
Comment (javiercn) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewComponents/ServiceBasedViewComponentActivator.cs]: Our service provider doesn't do anything on this method, so I'm not sure we want to add that. I'm sticking to the /// <inheritdoc /> comment here.

**Comment by javiercn** [comment]
Updated, cleaned up. This is not design anymore

File: src/Microsoft.AspNetCore.Mvc.ViewFeatures/DependencyInjection/MvcViewFeaturesMvcBuilderExtensions.cs [hunk 8]
```diff
@@ -33,5 +37,30 @@ public static class MvcViewFeaturesMvcBuilderExtensions
             builder.Services.Configure(setupAction);
             return builder;
         }
+
+        /// <summary>
+        /// Registers discovered view components as services in the <see cref="IServiceCollection"/>
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.ViewFeatures/DependencyInjection/MvcViewFeaturesMvcBuilderExtensions.cs]: `.`

File: test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentSelectorTest.cs [hunk 9]
```diff
@@ -188,39 +188,48 @@ public class FullNameInAttribute
...
+                {
+                    foreach (var type in parts.OfType<IApplicationPartTypeProvider>().SelectMany(p => p.Types))
+                    {
+                        feature.ViewComponents.Add(type);
+                    }
+                }
             }
 
-            private static IAssemblyProvider GetAssemblyProvider()
+            private class TestPart : ApplicationPart, IApplicationPartTypeProvider
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.ViewFeatures.Test/ViewComponents/DefaultViewComponentSelectorTest.cs]: Should we just make this a public class? It's duplicated 3-4 times that I've seen so far.

**Comment by rynowak** [comment]
Can you add a functional test? Reuse the `ControllersAsServices` project

:watch:

**Comment by javiercn** [comment]
Updated

File: test/Microsoft.AspNetCore.Mvc.FunctionalTests/ViewComponentFromServicesTests.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            // Act
+            var response = await Client.SendAsync(request);
+            var responseText = await response.Content.ReadAsStringAsync();
+
+            // Assert
+            Assert.Equal(expected, responseText);
+        }
+
+        [Fact]
+        public async Task NonRegisteredViewComponentReturns500()
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.FunctionalTests/ViewComponentFromServicesTests.cs]: Remove this. We don't want to create functional tests that throw

**Comment by rynowak** [comment]
:shipit:

**Pull request closed without merging**